ceph/rados/core_workflows.py

```diff
@@ -2021,3 +2021,159 @@ def check_pool_pg_states(self, pool: str, disallowed_states: list):
+        cmd_get_obj_key = (
+            f"{cmd_base} ceph-objectstore-tool --data-path /var/lib/ceph/osd/ceph-"
+            f"{osd_id} --pgid {pg_id} {object_name}  list-omap"
+        )
+        out_put = acting_osd_node.exec_command(sudo=True, cmd=cmd_get_obj_key)
+        key_list = list(filter(None, out_put[0].split("\n")))
+        return key_list
+
+    def rm_object_key(self, osd_id, pg_id, object_name, object_key):
+        """
+        Method removes the object key
+        Args:
+            osd_id: osd id number
+            pg_id: pg id number
+            object_name: object name
+            object_key: object key to remove that mapped to the object
+
+        Returns: True -> Key deleted False -> Key not deleted
+        """
+        cmd_base = f"cephadm shell --name osd.{osd_id} --"
+        acting_osd_node = self.fetch_host_node(daemon_type="osd", daemon_id=osd_id)
+        cmd_rm_obj_key = (
+            f"{cmd_base} ceph-objectstore-tool --data-path /var/lib/ceph/osd/ceph-"
+            f"{osd_id} --pgid {pg_id} {object_name}  rm-omap {object_key}"
+        )
+        try:
+            acting_osd_node.exec_command(sudo=True, cmd=cmd_rm_obj_key)
+        except Exception:
+            log.error(
+                f"{object_key} object key is not removed for the {object_name} object"
+            )
+            return False
+        log.info(f"{object_key} object key is removed for the {object_name} object")
+        return True
+
+    def get_inconsistent_pg_list(self, pool_name):
+        """
+        Method returns the inconsistent pg list
+        Args:
+            pool_name:  pool name
+
+        Returns: Inconsistent pg list
+        """
+        cmd_inconsist_pg = f"rados list-inconsistent-pg  {pool_name}"
+        inconsistent_pg_list = self.run_ceph_command(cmd_inconsist_pg)
+        if not inconsistent_pg_list:
+            log.info(f"The inconsistent pg list is empty in the pool {pool_name}")
+        return inconsistent_pg_list
+
+    def get_inconsistent_object_details(self, pg_id):
+        """
+        Method returns the inconsistent object list from a PG
+        Args:
+            pg_id: pg id
+
+        Returns: Return the inconsistent object list
+        """
+        cmd_inconsistent_object = f"rados  list-inconsistent-obj  {pg_id}"
+        object_details = self.run_ceph_command(cmd_inconsistent_object)
+        if not object_details:
+            log.info(f"The inconsistent objects list is empty in the PG {pg_id}")
+        return object_details
+
+    def check_inconsistent_health(self):
+        """
+        Method perform the health check for the inconsistent objects
+        Returns: True -> Contain inconsistent objects
+                 False -> Not contain inconsistent objects
+        """
+        health_detail, _ = self.node.shell(args=["ceph health detail"])
+        log.info(f"Health warning: \n {health_detail}")
+        if "inconsistent*" not in health_detail:
```

Comment:
@SrinivasaBharath Do we need "*" in if condition here? From the logs I can see inconsistent being generated in health warning however we are logging as "pg inconsistent not generated in cluster" 

Log snippet:
2023-10-16 16:33:33,375 (cephci.test_osd_inconsistency_pg) [INFO] - cephci.ceph.rados.core_workflows.py:2127 - Health warning: 
 HEALTH_ERR Failed to place 1 daemon(s); 2 scrub errors; Possible data damage: 1 pg inconsistent
[WRN] CEPHADM_DAEMON_PLACE_FAIL: Failed to place 1 daemon(s)
    Failed while placing grafana.ceph-omap-b1cnxk-node6 on ceph-omap-b1cnxk-node6: cephadm exited with an error code: 1, stderr: Non-zero exit code 125 from /usr/bin/podman container inspect --format {{.State.Status}} ceph-6f550a60-6c60-11ee-a23a-fa163eefcacc-grafana-ceph-omap-b1cnxk-node6
/usr/bin/podman: stderr Error: inspecting object: no such container ceph-6f550a60-6c60-11ee-a23a-fa163eefcacc-grafana-ceph-omap-b1cnxk-node6
Non-zero exit code 125 from /usr/bin/podman container inspect --format {{.State.Status}} ceph-6f550a60-6c60-11ee-a23a-fa163eefcacc-grafana.ceph-omap-b1cnxk-node6
/usr/bin/podman: stderr Error: inspecting object: no such container ceph-6f550a60-6c60-11ee-a23a-fa163eefcacc-grafana.ceph-omap-b1cnxk-node6
Deploy daemon grafana.ceph-omap-b1cnxk-node6 ...
Non-zero exit code 125 from /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --authfile=/etc/ceph/podman-auth.json --net=host --entrypoint stat --init -e CONTAINER_IMAGE=registry.redhat.io/rhceph/grafana-rhel9:latest -e NODE_NAME=ceph-omap-b1cnxk-node6 -e CEPH_USE_RANDOM_NONCE=1 registry.redhat.io/rhceph/grafana-rhel9:latest -c %u %g /var/lib/grafana
stat: stderr Trying to pull registry.redhat.io/rhceph/grafana-rhel9:latest...
stat: stderr Error: initializing source docker://registry.redhat.io/rhceph/grafana-rhel9:latest: reading manifest latest in registry.redhat.io/rhceph/grafana-rhel9: unauthorized: access to the requested resource is not authorized
ERROR: Failed to extract uid/gid for path /var/lib/grafana: Failed command: /usr/bin/podman run --rm --ipc=host --stop-signal=SIGTERM --authfile=/etc/ceph/podman-auth.json --net=host --entrypoint stat --init -e CONTAINER_IMAGE=registry.redhat.io/rhceph/grafana-rhel9:latest -e NODE_NAME=ceph-omap-b1cnxk-node6 -e CEPH_USE_RANDOM_NONCE=1 registry.redhat.io/rhceph/grafana-rhel9:latest -c %u %g /var/lib/grafana: Trying to pull registry.redhat.io/rhceph/grafana-rhel9:latest...
Error: initializing source docker://registry.redhat.io/rhceph/grafana-rhel9:latest: reading manifest latest in registry.redhat.io/rhceph/grafana-rhel9: unauthorized: access to the requested resource is not authorized
[ERR] OSD_SCRUB_ERRORS: 2 scrub errors
[ERR] PG_DAMAGED: Possible data damage: 1 pg inconsistent
    pg 8.4 is active+clean+inconsistent, acting [0,2,10]

2023-10-16 16:33:33,375 (cephci.test_osd_inconsistency_pg) [INFO] - cephci.ceph.rados.core_workflows.py:2129 - pg inconsistent not generated in the cluster

Reply:
Good catch Neha!
I fix the code. Test execution is in progress. I will share the logs once it is done.

Reply:
sure thanks, one pass log is sufficient. Looks good overall.

Reply:
Reef pass log- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-4GFCNP